### DISCUSSION TITLE
How can I add onKeyPressed/onKeyReleased functionality to a button?

### DISCUSSION THREAD

**Discussion by DAT4** [fyne-io/fyne#3396] [Q&A] [answered]
It seems like if should be easy to implement - But I can't see any direct way without having to implement my own button from scratch.

**Comment by Jacalz** [comment, marked as answer]
You basically want to follow the tutorials on https://developer.fyne.io/extend/extending-widgets and https://developer.fyne.io/extend/numerical-entry to get an idea about how to extend widgets. It might also be good to watch https://www.youtube.com/watch?v=TDae1Kmxu_0 if you want the full understanding of how the interfaces work when you want to add additional functionality to your widgets.

To actually get the functionality you request, it would require implementing [desktop.Keyable](https://pkg.go.dev/fyne.io/fyne/v2/driver/desktop#Keyable) ([desktop.Mouseable](https://pkg.go.dev/fyne.io/fyne/v2/driver/desktop#Mouseable) for mouse keys for the record but I don't think that is what you are asking for) to be able to get that kind of funtionality.

**Reply by DAT4** [reply]
Thanks you @Jacalz  - I followed your advice and implemented my own button

```go
package components

import (
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/driver/desktop"
	"fyne.io/fyne/v2/widget"
)

type MyButton struct {
	widget.Button
	fyne.Focusable
	OnKeyDown func() `json:"-"`
	OnKeyUp   func() `json:"-"`
}

func NewMyButton(title string, keyDown, keyUp func()) *MyButton {
	return &MyButton{
		Button:    *widget.NewButton(title, nil),
		OnKeyUp:   keyUp,
		OnKeyDown: keyDown,
	}
}

func (b *MyButton) handleClickEvent(fn func()) {
	if b.Disabled() {
		return
	}

	b.Refresh()

	if fn != nil {
		fn()
	}
}

func (b *MyButton) MouseDown(*desktop.MouseEvent) {
	b.handleClickEvent(b.OnKeyDown)
}

func (b *MyButton) MouseUp(*desktop.MouseEvent) {
	b.handleClickEvent(b.OnKeyUp)
}

func (b *MyButton) KeyDown(*fyne.KeyEvent) {
	b.handleClickEvent(b.OnKeyDown)
}

func (b *MyButton) KeyUp(*fyne.KeyEvent) {
	b.handleClickEvent(b.OnKeyUp)
}
```

This works in a desktop application - however when running it on android the touch is not registered - So I guess I have to find a Touchable interface and implement it :)

**Reply by andydotxyz** [reply]
You should not have to re-implement the tap handling. It looks like you forgot to call `ExtendBaseWidget`
